### PR TITLE
Include location footer in logs for AssertionErrors

### PR DESCRIPTION
**Pull request by jayeshathila**
Changes:
```
1. Adding Location footer in AssertionErrors.
```

Concerns:
```
1. Didn't add tests for the change as I can't find other tests for the Locations.
2. Approach doesn't seem clean to me, not sure if how we can improve it.
```
[issue #126 ](https://github.com/darrenburns/ward/issues/126)

### CODE REVIEW

**Comment by darrenburns** [comment]
I wonder if there's a way we could have the location always be the line number that the test failed on, just to keep things consistent throughout? If so, I also think we should print out the surrounding context of where the failure occurred, as per the "Failure" header shown in the image below.

Screenshot 2020-02-04 at 22 26 52

**Comment by blueyed** [comment]
The location (file + lnum) of the test failure is very useful, since it allows to quickly go there in an editor.
Keep in mind though that it might fail before or after the actual test however (fixture setup/teardown etc).  In that case it should be the location of the "crash" probably.

**Comment by darrenburns** [comment]
Very good point @blueyed. 

There's two relevant lines: the location of the failing test definition, and the location the failure/error actually occurs. I wonder what the least confusing way to communicate those to the user is? Any thoughts @jayeshathila? I'd be open to changing the existing output if we can think of something clearer.

**Comment by blueyed** [comment]
With pytest I am using `nodeid::test_file.py:42` by default, but expand it to `nodeid::test_file.py:42 (some/other/file.py:123)` if the crash happened somewhere else.  This provides two locations then: one to easily go to the test's code that triggered it, and another for going to the crash itself (which might be a lib or a non-hidden test helper).

**Comment by jayeshathila** [comment]
> Very good point @blueyed.
> 
> There's two relevant lines: the location of the failing test definition, and the location the failure/error actually occurs. I wonder what the least confusing way to communicate those to the user is? Any thoughts @jayeshathila? I'd be open to changing the existing output if we can think of something clearer.

Ya, I am myself aligned with that we should print the line_num where failure actually occur. Will see what can be done to print the line_num. Will update my finding here.

**Comment by jayeshathila** [comment]
@darrenburns @blueyed I have added line_number explicitly (using `traceback`)  in case of AssertionError while creating TestResult [here](https://github.com/darrenburns/ward/pull/128/commits/a58489ac885294bb89183b1019c07214b344f450#diff-2aa84a372e3f58608667709b8925e6e6R222). Any thoughts on the approach ?

File: ward/testing.py [hunk 1]
```diff
@@ -217,6 +219,12 @@ def get_result(self, outcome, exception=None):
             if outcome in (TestOutcome.PASS, TestOutcome.SKIP):
                 result = TestResult(self, outcome)
             else:
+                if isinstance(exception, AssertionError):
+                    exc_traceback = sys.exc_info()[-1]
```
Comment (darrenburns) [inline ward/testing.py]: My understanding is that `sys.exc_info` only works in the context of `except` blocks?

I think you can get the traceback from the exception using `exc_traceback = exception.__traceback__`.

Comment (jayeshathila) [reply ward/testing.py]: Ah. Agree `get_result` being public method I shouldn't expect it to be in context of `except`. Will change this to use `exception.__traceback`

Comment (jayeshathila) [reply ward/testing.py]: > My understanding is that `sys.exc_info` only works in the context of `except` blocks?
> 
> I think you can get the traceback from the exception using `exc_traceback = exception.__traceback__`.

Ah. Agree get_result being public method I shouldn't expect it to be in context of except. Will change this to use exception.__traceback

Comment (jayeshathila) [reply ward/testing.py]: @darrenburns, Updated with `exception.__traceback__`.

**Comment by darrenburns** [comment]
Looks good, thanks @jayeshathila. Will take a closer look asap!

**Comment by darrenburns** [comment]
This change has been included in Ward 0.39.0b0 - thanks again @jayeshathila :)

**Pull request merged**